internal/collector/gpu/gpu.go

```diff
@@ -0,0 +1,398 @@
+
+import (
+	"fmt"
+	"log/slog"
+	"sync"
+
+	"github.com/alecthomas/kingpin/v2"
+	"github.com/prometheus-community/windows_exporter/internal/mi"
+	"github.com/prometheus-community/windows_exporter/internal/pdh"
+	"github.com/prometheus-community/windows_exporter/internal/types"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const Name = "gpu"
+
+type Config struct{}
+
+var ConfigDefaults = Config{}
+
+type Collector struct {
+	config Config
+
+	// GPU Engine
+	gpuEnginePerfDataCollector *pdh.Collector
+	gpuEnginePerfDataObject    []gpuEnginePerfDataCounterValues
+
+	gpuEngineRunningTime           *prometheus.Desc
+	gpuEngineUtilizationPercentage *prometheus.Desc
+
+	// GPU Adapter Memory
+	gpuAdapterMemoryPerfDataCollector *pdh.Collector
+	gpuAdapterMemoryPerfDataObject    []gpuAdapterMemoryPerfDataCounterValues
+
+	gpuAdapterMemoryDedicatedUsage *prometheus.Desc
+	gpuAdapterMemorySharedUsage    *prometheus.Desc
+	gpuAdapterMemoryTotalCommitted *prometheus.Desc
+
+	// GPU Local Adapter Memory
+	gpuLocalAdapterMemoryPerfDataCollector *pdh.Collector
+	gpuLocalAdapterMemoryPerfDataObject    []gpuLocalAdapterMemoryPerfDataCounterValues
+
+	gpuLocalAdapterMemoryUsage *prometheus.Desc
+
+	// GPU Non Local Adapter Memory
+	gpuNonLocalAdapterMemoryPerfDataCollector *pdh.Collector
+	gpuNonLocalAdapterMemoryPerfDataObject    []gpuNonLocalAdapterMemoryPerfDataCounterValues
+
+	gpuNonLocalAdapterMemoryUsage *prometheus.Desc
+
+	// GPU Process Memory
+	gpuProcessMemoryPerfDataCollector *pdh.Collector
+	gpuProcessMemoryPerfDataObject    []gpuProcessMemoryPerfDataCounterValues
+
+	gpuProcessMemoryDedicatedUsage *prometheus.Desc
+	gpuProcessMemoryLocalUsage     *prometheus.Desc
+	gpuProcessMemoryNonLocalUsage  *prometheus.Desc
+	gpuProcessMemorySharedUsage    *prometheus.Desc
+	gpuProcessMemoryTotalCommitted *prometheus.Desc
+
+	mu sync.Mutex
+}
+
+func New(config *Config) *Collector {
+	if config == nil {
+		config = &ConfigDefaults
+	}
+
+	c := &Collector{
+		config: *config,
+	}
+
+	return c
+}
+
+func NewWithFlags(_ *kingpin.Application) *Collector {
+	return &Collector{}
+}
+
+func (c *Collector) GetName() string {
+	return Name
+}
+
+func (c *Collector) Close() error {
+	c.gpuEnginePerfDataCollector.Close()
+	c.gpuAdapterMemoryPerfDataCollector.Close()
+	c.gpuLocalAdapterMemoryPerfDataCollector.Close()
+	c.gpuNonLocalAdapterMemoryPerfDataCollector.Close()
+	c.gpuProcessMemoryPerfDataCollector.Close()
+
+	return nil
+}
+
+func (c *Collector) Build(_ *slog.Logger, _ *mi.Session) error {
+	var err error
+
+	c.gpuEngineRunningTime = prometheus.NewDesc(
+		prometheus.BuildFQName(types.Namespace, Name, "engine_time_seconds"),
+		"Total running time of the GPU in seconds.",
+		[]string{"pid", "phys", "eng", "engtype"},
+		nil,
+	)
+	c.gpuEngineUtilizationPercentage = prometheus.NewDesc(
+		prometheus.BuildFQName(types.Namespace, Name, "utilization_percent"),
+		"GPU utilization percentage.",
+		[]string{"pid", "phys", "eng", "engtype"},
+		nil,
+	)
```

Comment:
Normally, its sufficient to expose the engine_time_seconds to calculation the utilization_percent. At least this is done at CPU.